src/Microsoft.VisualStudio.Threading/JoinableTask`1.cs

```diff
@@ -47,10 +47,27 @@ internal JoinableTask(JoinableTaskFactory owner, bool synchronouslyBlocking, Joi
         /// </summary>
         /// <param name="cancellationToken">A cancellation token that will exit this method before the task is completed.</param>
         /// <returns>A task that completes after the asynchronous operation completes and the join is reverted, with the result of the operation.</returns>
-        public new async Task<T> JoinAsync(CancellationToken cancellationToken = default(CancellationToken))
+        public new Task<T> JoinAsync(CancellationToken cancellationToken = default(CancellationToken))
         {
-            await base.JoinAsync(cancellationToken).ConfigureAwait(AwaitShouldCaptureSyncContext);
-            return await this.Task.ConfigureAwait(AwaitShouldCaptureSyncContext);
+            Task baseJoin = base.JoinAsync(cancellationToken);
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return System.Threading.Tasks.Task.FromCanceled<T>(cancellationToken);
+            }
+
+            if (baseJoin.IsCompleted)
+            {
+                Assumes.True(this.Task.IsCompleted);
+                return this.Task;
+            }
+
+            return JoinSlowAsync(baseJoin);
+
+            async Task<T> JoinSlowAsync(Task baseJoin)
+            {
+                await baseJoin.ConfigureAwait(AwaitShouldCaptureSyncContext);
+                return await this.Task.ConfigureAwait(AwaitShouldCaptureSyncContext);
```

Comment:
Doesn't `this` create a closure, making this nested method allocate two more objects just to invoke it than if we had defined this as its own `private` method?

Reply:
The closure was removed from the containing method, so this is the same allocation that already existed, just in a slightly different location.

Reply:
Can we avoid the closure though by making this a private method?

Reply:
The current form is the same as a private method. The closure occurs due to the use of `async`, not due to the fact that this is a local function.

https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYYCWAdgC7oCycAFAJT6HFtQDs6AMgPYDGAQwA2UAKyVSvcvQAMjANys2RZSorU+Q0RKkyN6AG4iArgFNmBJCpvoAEmeEAHMwCcmS67fZdjw857eAL5qIUhqaOgGNDBMLF42nLRmlAAWvMByiqHIak6upMaUZlFUyWkZ9AZ+5pZqbA7Obh71xEk1ZoHEYWx5BUUlUAAs9o4u7pboYUFAA===